sql: introduce ContainsAggregateClass for builtins

otan opened the pull request:
We formerly expected all overloads of an AggregateClass to only contain
aggregates. However, as is the case in PostGIS, this is not strictly
true. To account for this, add a new class `ContainsAggregateClass`,
which behaves mostly as AggregateClass except that it allows
non-aggregates to be included the the typename, hence showing up in the
correct places in docs and keeping the error messages for pure
AggregateClass functions.

This is demonstrated with ST_MakeLine, which has both an aggregate
version (existing and tested) and the non-aggregate version.

Resolves #51508
Resolves #48980

Release note (sql change): Implement the ST_MakeLine(a, b) variant,
which takes in two geometries and returns a line out of both.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52830)
<!-- Reviewable:end -->

rytaft reviewed (commented):
> Reviewed 12 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx90YZ8Lf_DNrZII-Y:-MEx90YZ8Lf_DNrZII-Z:b1qeki5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/logictest/testdata/logic_test/geospatial#L4039)):*
> ```
> 
> query T
> SELECT ST_AsEWKT(ST_MakeLine(ST_MakePoint(1, 2), ST_MakePoint(3, 4)))
> ```

How about adding some tests to exercise some of the weird type checking stuff? E.g., what happens if you pass NULL arguments or use a FILTER clause? How about nesting an aggregate function inside ST_MakeLine or vice-versa?
___
*[pkg/sql/opt/optbuilder/scope.go, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MExS2Q18SpiwQbIJ3Ew:-MExS2Q2AFHf812n7X9r:bqe6aa9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/opt/optbuilder/scope.go#L922)):*
> ```Go
> 				expr = replacedExpr
> 			} else if isAggregate(def) {
> 				panic(errors.Newf("expected %s to be replaced with an aggregate", def.Name))
> ```

Maybe this should be `AssertionFailedf`?
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 1343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx7-bjDeU1YW5EZV5o:-MEx7-bkCfOik_SDnA1b:b-ovrc9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/sem/builtins/geo_builtins.go#L1343)):*
> ```Go
> 			},
> 			Info: infoBuilder{
> 				info: "Forms a LineString from Point, MultiPoint or LineStrings. Other shapes will be ignored.",
> ```

[nit] Forms a LineString from two Points, MultiPoints, or LineStrings...
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 1356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx7En5DFBXz6YouJVq:-MEx7En5DFBXz6YouJVr:beuoaqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/sem/builtins/geo_builtins.go#L1356)):*
> ```Go
> 			},
> 			infoBuilder{
> 				info: "Forms a LineString from Point, MultiPoint or LineStrings. Other shapes will be ignored.",
> ```

[nit] Forms a LineString from Points, MultiPoints, or LineStrings...


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx90YZ8Lf_DNrZII-Y:-MExaKjW9LALQoSwcjPl:b-np6jrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/logictest/testdata/logic_test/geospatial#L4039)):*
> what happens if you pass NULL arguments or use a FILTER clause?
added an extra one

> How about nesting an aggregate function inside ST_MakeLine or vice-versa?
done!
___
*[pkg/sql/opt/optbuilder/scope.go, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MExS2Q18SpiwQbIJ3Ew:-MExaFW4BN1WHUkkqtcR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/opt/optbuilder/scope.go#L922)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe this should be `AssertionFailedf`?
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 1343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx7-bjDeU1YW5EZV5o:-MExaGQgCrvmdlmX60FV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/sem/builtins/geo_builtins.go#L1343)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] Forms a LineString from two Points, MultiPoints, or LineStrings...
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 1356 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx7En5DFBXz6YouJVq:-MExaGfI9pwm5DmzgPMJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/sem/builtins/geo_builtins.go#L1356)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] Forms a LineString from Points, MultiPoints, or LineStrings...
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
sql: introduce ContainsAggregateClass for builtins

We formerly expected all overloads of an AggregateClass to only contain
aggregates. However, as is the case in PostGIS, this is not strictly
true. To account for this, add a new class `ContainsAggregateClass`,
which behaves mostly as AggregateClass except that it allows
non-aggregates to be included the the typename, hence showing up in the
correct places in docs and keeping the error messages for pure
AggregateClass functions.

This is demonstrated with ST_MakeLine, which has both an aggregate
version (existing and tested) and the non-aggregate version.

Release note (sql change): Implement the ST_MakeLine(a, b) variant,
which takes in two geometries and returns a line out of both.

rytaft reviewed (commented):
> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx90YZ8Lf_DNrZII-Y:-MExmFBA5F04t9vUjC0N:b-oubwbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/logictest/testdata/logic_test/geospatial#L4039)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> what happens if you pass NULL arguments or use a FILTER clause?
added an extra one

> How about nesting an aggregate function inside ST_MakeLine or vice-versa?
done!
</blockquote></details>

Nice. Would be good to check a couple of error cases too, like using `FILTER` with the non-aggregate version of `ST_MakeLine` and nesting an aggregate like `ST_MakeLine(ST_MakeLine(g::geometry))`
___
*[pkg/sql/opt/optbuilder/scope.go, line 1126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MExneCA9aqHwfKe40kB:-MExneCB7rCdcX7XBhwW:b-weuitr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8613ba39999b77ad6337f2a81bc768d5c8c10b75/pkg/sql/opt/optbuilder/scope.go#L1126)):*
> ```Go
> 	// return an appropriate error. The returned tempScope will be used for
> 	// building aggregate function arguments below in buildAggregateFunction.
> 	tempScope := s.startAggFunc()
> ```

I think this has to happen before the calls to TypeCheck above -- moving this is probably going to cause some other tests to fail...


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx90YZ8Lf_DNrZII-Y:-MExxzWFC23jmzYUVnQa:b-t4c24q) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/logictest/testdata/logic_test/geospatial#L4039)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Nice. Would be good to check a couple of error cases too, like using `FILTER` with the non-aggregate version of `ST_MakeLine` and nesting an aggregate like `ST_MakeLine(ST_MakeLine(g::geometry))`
</blockquote></details>

I'm unable to make both the ST_MakeLine(ST_MakeLine(g::geometry)) and ST_MakeLine(agg, non_agg) work :\

same with the FILTER item -- i need the resolved overload first before applying checks like these.

I'm thinking we probably shouldn't go through with this without a proper restructuring of type checking, not sure we'll get to that this release.

I'd say put this on the backburner for now (we don't know how urgent it is)


<!-- Sent from Reviewable.io -->

otan commented:
throwing this work away for now -- need to do a lot of type check restructures to get the same kind of checks that we expect to work.

rytaft commented:
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 4039 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52830#-MEx90YZ8Lf_DNrZII-Y:-MEy198m7YJFKRFHxOfN:b-pyvegm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fa7f3830b65c6093f2960d39c48caa89dd0e03f/pkg/sql/logictest/testdata/logic_test/geospatial#L4039)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I'm unable to make both the ST_MakeLine(ST_MakeLine(g::geometry)) and ST_MakeLine(agg, non_agg) work :\

same with the FILTER item -- i need the resolved overload first before applying checks like these.

I'm thinking we probably shouldn't go through with this without a proper restructuring of type checking, not sure we'll get to that this release.

I'd say put this on the backburner for now (we don't know how urgent it is)
</blockquote></details>

Oh well :woman_shrugging: Makes sense


<!-- Sent from Reviewable.io -->
